clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -168,76 +329,686 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
         }
-        return state.equals(MemberState.STABLE);
+        return reason;
+    }
+
+    /**
+     *  Visible for testing.
+     */
+    void markReconciliationInProgress() {
+        reconciliationInProgress = true;
+        memberEpochOnReconciliationStart = memberEpoch;
+    }
+
+    /**
+     *  Visible for testing.
+     */
+    void markReconciliationCompleted() {
+        reconciliationInProgress = false;
     }
 
     /**
-     * Take new target assignment received from the server and set it as targetAssignment to be
-     * processed. Following the consumer group protocol, the server won't send a new target
-     * member while a previous one hasn't been acknowledged by the member, so this will fail
-     * if a target assignment already exists.
+     * Build set of TopicPartition (topic name and partition id) from the target assignment
+     * received from the broker (topic IDs and list of partitions).
+     *
+     * <p>
+     * This will:
      *
-     * @throws IllegalStateException If a target assignment already exists.
+     * <ol type="1">
+     *     <li>Try to find topic names in the metadata cache</li>
+     *     <li>For topics not found in metadata, try to find names in the local topic names cache
+     *     (contains topic id and names currently assigned and resolved)</li>
+     *     <li>If there are topics that are not in metadata cache or in the local cached
+     *     of topic names assigned to this member, request a metadata update, and continue
+     *     resolving names as the cache is updated.
+     *     </li>
+     * </ol>
      */
-    private void setTargetAssignment(ConsumerGroupHeartbeatResponseData.Assignment newTargetAssignment) {
-        if (!targetAssignment.isPresent()) {
-            log.info("Member {} accepted new target assignment {} to reconcile", memberId, newTargetAssignment);
-            targetAssignment = Optional.of(newTargetAssignment);
+    private void resolveMetadataForUnresolvedAssignment() {
+        // Try to resolve topic names from metadata cache or subscription cache, and move
+        // assignments from the "unresolved" collection, to the "readyToReconcile" one.
+        Iterator<Map.Entry<Uuid, List<Integer>>> it = assignmentUnresolved.entrySet().iterator();
+        while (it.hasNext()) {
+            Map.Entry<Uuid, List<Integer>> e = it.next();
+            Uuid topicId = e.getKey();
+            List<Integer> topicPartitions = e.getValue();
+
+            Optional<String> nameFromMetadata = findTopicNameInGlobalOrLocalCache(topicId);
+            nameFromMetadata.ifPresent(resolvedTopicName -> {
+                // Name resolved, so assignment is ready for reconciliation.
+                SortedSet<TopicIdPartition> topicIdPartitions =
+                        buildAssignedPartitionsWithTopicName(topicId, resolvedTopicName, topicPartitions);
+                assignmentReadyToReconcile.addAll(topicIdPartitions);
```

Comment:
I am curious here. Is it better to build a `SortedSet` with all elements and then to add it to `assignmentReadyToReconcile` vs adding to `assignmentReadyToReconcile` directly?

Reply:
Yeah, no value in it. I simplified it by just adding the TopicPartition items directly to the `assignmentReadyToReconcile` (in [follow-up PR](https://github.com/apache/kafka/pull/14857))